metadatamaker/util.go

```diff
@@ -0,0 +1,170 @@
+package metadatamaker
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+)
+
+// https://golang.org/ref/spec#Slice_types
+// max len is int-1 which is size of the default integer on target build
+// TODO find out if we need error handling here?
+// we might want to do system sanity check at compile time
+// or possibly use something other...
+// this is kind of edge case either way
+func pushdata(buf []byte) []byte {
+	bufLen := len(buf)
+
+	if bufLen == 0 {
+		return []byte{0x4C, 0x00}
+	} else if bufLen < 0x4E {
+		return bytes.Join([][]byte{[]byte{uint8(bufLen)}, buf}, []byte{})
+	} else if bufLen < 0xFF {
+		return bytes.Join([][]byte{[]byte{0x4C, uint8(bufLen)}, buf}, []byte{})
+	} else if bufLen < 0xFFFF {
+		tmp := make([]byte, 2)
+		binary.LittleEndian.PutUint16(tmp, uint16(bufLen))
+		return bytes.Join([][]byte{[]byte{0x4D}, tmp, buf}, []byte{})
+	} else if bufLen < 0xFFFFFFFF {
+		tmp := make([]byte, 4)
+		binary.LittleEndian.PutUint32(tmp, uint32(bufLen))
+		return bytes.Join([][]byte{[]byte{0x4E}, tmp, buf}, []byte{})
+	} else {
+		tmp := make([]byte, 8)
+		binary.LittleEndian.PutUint64(tmp, uint64(bufLen))
+		return bytes.Join([][]byte{[]byte{0x4F}, tmp, buf}, []byte{})
+	}
+}
+
+// TODO we can use better name for this
+func makeU64BigEndianBytes(v uint64) []byte {
+	tmp := make([]byte, 8)
+	binary.BigEndian.PutUint64(tmp, v)
+	return tmp
+}
+
+// mintBatonVout -- 0 means null
+// TODO is there more idiomatic way to represent optional here?
```

Comment:
Not with basic types like an int. What you can do is wrap it in another object and pass a pointer, then you can get a nil check.

Reply:
Thanks, done in follow up PR